Fix #1321 and #1459 and #171 and #1320 and #1456

Elv13 opened the pull request:
Before: to everything twice including the request::geometry
Now: do it once

Edit: also create a real `maximized` property, spliting the state across 2 properties is a source of corner cases.

@taptap @cagelight can you test?

Fix #1320 #1321 #171 #1459

psychon commented on objects/client.h:
```diff
@@ -47,6 +47,12 @@ typedef enum {
+    CLIENT_MAXIMIZED_NONE = 0 << 0,
+    CLIENT_MAXIMIZED_V    = 1 << 0,
+    CLIENT_MAXIMIZED_H    = 1 << 1,
+} client_maximized_t;
```
Not needed in the header. Can't this be right in the C file?

Elv13 replied:
Done

Elv13 commented:
Ah, as I think of it, its a behavior change for the 0.1 users who set both manually. I should change this to be additive instead of overwriting.

psychon commented on objects/client.c:
```diff
@@ -1919,38 +1920,68 @@ client_get_maximized(client_t *c)
+    const client_maximized_t current = (
+        (c->maximized_vertical   ? CLIENT_MAXIMIZED_V : CLIENT_MAXIMIZED_NONE)|
+        (c->maximized_horizontal ? CLIENT_MAXIMIZED_H : CLIENT_MAXIMIZED_NONE)
+    );
```
Well, ok. One thing: Why two bits instead of just using two different variables for this? GCC can likely figure out nice and fast code anyway (not that it matters much here...) and I think the result would be more readable.

Elv13 replied:
The property macro like booleans

taptap commented:
It still restores on the monitor where it was maximized.
However, it's not widened to whole width anymore.

Bew78LesellB commented:
It works well for me, this fix #1459, thanks !

psychon commented on objects/client.c:
```diff
@@ -1919,38 +1926,68 @@ client_get_maximized(client_t *c)
+        (c->maximized_vertical   ? CLIENT_MAXIMIZED_V : CLIENT_MAXIMIZED_NONE)|
+        (c->maximized_horizontal ? CLIENT_MAXIMIZED_H : CLIENT_MAXIMIZED_NONE)
+    );
+    const client_maximized_t next = s ? val : CLIENT_MAXIMIZED_NONE;
```
Likely makes no difference, but this could be moved into the caller. Instead of having a boolean and a mask, this function could directly get the `next` value.
</bikeshedding>

psychon commented on tests/test-maximize.lua:
```diff
@@ -99,6 +100,47 @@ local steps = {
+
+        if c.maximized then
+            return
+        end
```
Is this `if` necessary? Setting `c.maximized` should be instantaneous (so this could even be done in the previous step)

psychon commented on ewmh.c:
```diff
@@ -216,6 +216,7 @@ ewmh_init(void)
     luaA_class_connect_signal(L, &client_class, "property::fullscreen" , ewmh_client_update_hints);
     luaA_class_connect_signal(L, &client_class, "property::maximized_horizontal" , ewmh_client_update_hints);
     luaA_class_connect_signal(L, &client_class, "property::maximized_vertical" , ewmh_client_update_hints);
+    luaA_class_connect_signal(L, &client_class, "property::maximized" , ewmh_client_update_hints);
```
So... a client can be maximized vertically, maximized horizontally and at the same time not be maximized? Put differently, these are now three completely independent booleans? And all code that wants to check for `maximized_vertical` should now check `maximized_vertical or maximized`, because the C code no longer does so for you? That seems rather odd and I don't like it at all.

How about deprecating the horizontal/vertical maximization instead and only have a single "maximized" state instead?

Elv13 replied:
The previous version of the patch kept the 2bit==maximized concept, but it doesn't work after all.

*problem 1:*
When you maximized something, all 3 signals were sent. If some code handled "maximized" a "maximized_horizontally" (or all 3), then it was executed 3 times.

*problem 1.1:*
`request::geometry` was sent twice (before this PR) when you did a single operation. This is why it failed to serialize the size.

*problem 2:*
If you had the 3 maximize button in the titlebar, than where you toggled maximize twice, it would override the other 2 and do->undo would not restore the state as it was.

So, having a split state made most operation *more* completed than having 3 independent states. Also, have 3 booleans ensure that when changing one of them, only 1 signal is emitted. The previous code worked, I am not denying that, but it had the "design quirks build-in" while the request API blueyed and me implemented is designed to unify "concepts" into a single handler. It *could* be fixed in the handler, but I would rather have a simplified pipeline. This is what this patch does. Most users use maximize and not the other 2, so simplifying that one while making the other 2 a little more complex doesn't seem such a bad idea.

> How about deprecating the horizontal/vertical maximization instead and only have a single "maximized" state instead?

I don't think its necessary. Once this patch becomes stable, it wont need to be changed. I probably should merge `fullscreen` into this function as well. I will if I ever attack #758 and [insert z-index limitations issue I can't find right now].

psychon commented on lib/awful/placement.lua:
```diff
@@ -1422,6 +1422,11 @@ function placement.restore(d, args)
+    -- Apparently, 0x0 did happen to at least 1 person. Officially can't
+    -- happen, but can't be too careful.
+    if memento.width  and memento.width < 1 then return end
+    if memento.height and memento.height < 1 then return end
```
How about moving this into the creation of a memento instead, so that such an invalid memento cannot be created at all?

However, I am curious how this could happen. This seems to require `c:geometry()` to return 0 for width/height. This can only happen while the client is still being managed, because there is a time frame while the value is still 0 before the values coming from the X11 server are applied. This can only happen during the following signals: `new`, `property::x`, `property::y`, `property::width`, `property::screen`, `property::window`.

An (in my opinion) even better proposal: Instead of working around this in Lua, change the C code so that it emits the signals later (except for `new`, but no one should use that one anyway)?
Basically:
```diff
diff --git a/objects/client.c b/objects/client.c
index ec4ab424b..9f6bd138b 100644
--- a/objects/client.c
+++ b/objects/client.c
@@ -1405,8 +1405,6 @@ client_manage(xcb_window_t w, xcb_get_geometry_reply_t *wgeom, xcb_get_window_at
      * (Else, reparent could cause an UnmapNotify) */
     xcb_change_window_attributes(globalconf.connection, w, XCB_CW_EVENT_MASK, select_input_val);
 
-    luaA_object_emit_signal(L, -1, "property::window", 0);
-
     /* The frame window gets the border, not the real client window */
     xcb_configure_window(globalconf.connection, w,
                          XCB_CONFIG_WINDOW_BORDER_WIDTH,
@@ -1429,15 +1427,16 @@ client_manage(xcb_window_t w, xcb_get_geometry_reply_t *wgeom, xcb_get_window_at
 
     /* Store initial geometry and emits signals so we inform that geometry have
      * been set. */
-#define HANDLE_GEOM(attr) \
-    c->geometry.attr = wgeom->attr; \
-    luaA_object_emit_signal(L, -1, "property::" #attr, 0);
-HANDLE_GEOM(x)
-HANDLE_GEOM(y)
-HANDLE_GEOM(width)
-HANDLE_GEOM(height)
-#undef HANDLE_GEOM
-
+    c->geometry.x = wgeom->x;
+    c->geometry.y = wgeom->y;
+    c->geometry.width = wgeom->width;
+    c->geometry.height = wgeom->height;
+
+    luaA_object_emit_signal(L, -1, "property::x", 0);
+    luaA_object_emit_signal(L, -1, "property::y", 0);
+    luaA_object_emit_signal(L, -1, "property::width", 0);
+    luaA_object_emit_signal(L, -1, "property::height", 0);
+    luaA_object_emit_signal(L, -1, "property::window", 0);
     luaA_object_emit_signal(L, -1, "property::geometry", 0);
 
     /* Set border width */
```

psychon commented on lib/awful/ewmh.lua:
```diff
@@ -270,6 +275,10 @@ function ewmh.geometry(c, context, hints)
 
+        if props.honor_padding == nil and props.honor_workarea then
+            props.honor_padding = beautiful.maximized_honor_padding ~= false
+        end
```
The commit message says that this is specific to maximisation, but I fail to see that in the code...?

psychon commented on lib/awful/ewmh.lua:
```diff
@@ -284,7 +279,6 @@ client.connect_signal("request::tag", ewmh.tag)
 client.connect_signal("request::urgent", ewmh.urgent)
 client.connect_signal("request::geometry", ewmh.geometry)
 client.connect_signal("property::border_width", geometry_change)
-client.connect_signal("property::geometry", geometry_change)
```
...sounds like the function name no longer matches its job (although I am not quite sure what that job is). Could you clean that up?

psychon commented on lib/awful/ewmh.lua:
```diff
@@ -42,18 +42,13 @@ local function geometry_change(window)
+        "fullscreen", "maximized", "maximized_vertical", "maximized_horizontal"
+    } do
+        if window[prop] then
+            window:emit_signal("request::geometry", prop, {})
```
Should the last argument somehow indicate that this is just re-applying the way things should be?

psychon commented on tests/test-maximize.lua:
```diff
@@ -82,10 +82,11 @@ local steps = {
+        assert(new_geo.x==sgeo.x)
+        assert(new_geo.y==sgeo.y)
+        assert(new_geo.x+new_geo.width+2*bw==sgeo.x+sgeo.width)
+        assert(new_geo.y+new_geo.height+2*bw==sgeo.y+sgeo.height)
```
Do I understand correctly that the code doing ` window:geometry(window.screen.geometry)` was actually wrong (since it ignored the border width) and the wrongness managed to end up in a test? Yuck.

Elv13 replied:
That code path wasn't of my making ;) (but the test is). Actually, after thinking about it, it was probably done on purpose. Done this way, you get "real" fullscreen. This was initially why I left it that way.  But now it conflicts with the request API, so it has to use the standard code path or these issues will keep piling up.

psychon replied:
...so, does that mean that the next issue will be the "fullscreen" is no longer fullscreen? That's not good. Fullscreen should hide the border (=set border width to zero).

Elv13 replied:
Done

Elv13 commented:
I addressed many comments and added the fix for #1320. I don't expect this to be approved without another round of changes, but at least this help the pile of open PRs I have right now.

Again, I want to point out that adding another maximized mode was made to simplify the code, even if it makes the C code is little bit uglier.

psychon commented on objects/client.c:
```diff
@@ -1962,21 +1953,28 @@ client_set_maximized_common(lua_State *L, int cidx, bool s, const char* type, co
+
+    /* When both are already set during startup, assume `maximized` is true*/
+    if (next == (CLIENT_MAXIMIZED_H|CLIENT_MAXIMIZED_V) && !globalconf.loop)
+        next = CLIENT_MAXIMIZED_BOTH;
```
Shouldn't this by `|=` instead of `=`?
Or should we invent our own `MAXIMIZED` state for EWMH and use that one here? If so, feel free to open a todo-issue assigned to me telling me to get rid of this again.

Elv13 replied:
> Shouldn't this by |= instead of =?

Nah, this is one of the possible "bad side effect" of this. If someone manually set both `maximized_[*]`, then it's interpreted as maximized on on restart. The opposite, if `=|` was used and the user had no shortcut for `maximized_[*]`, then he/she/it would be stuck with a triple maximized client with no way to turn it off. The "issue" itself already exist, but isn't "as bad" since the default maximized shortcut is currently "destructive" in the sense toggling it twice will reset both maximization. From now on, if one of the `maximize_*` is set and not the other, then it will be harder for the user to turn it off. The `=` just makes sure it doesn't happen because of this change.

At first I was thinking your idea of another maximized xprop was a good idea, but thinking about it again, I think a better course of action might be to reset fullscreen and maximization properties when restarting. It is a behavior change compared to previous version and someone, somewhere, will notice. However it simplifies the algorithm and users don't restart Awesome as often as they used to, so it wont be as noticeable as before. Anyway, when you lose your layouts and stack order, keeping maximization cause a confusing result.

psychon commented on lib/awful/ewmh.lua:
```diff
@@ -280,7 +284,18 @@ function ewmh.geometry(c, context, hints)
+        if original_context == "fullscreen" and beautiful.fullscreen_hide_border ~= false then
+            repair_geometry_lock = true
+            c.border_width = 0
+            repair_geometry_lock = false
```
shouldn't this be:
```
local original = repair_geometry_lock
repair_geometry_lock = true
c.border_width = 0
repair_geometry_lock = original
```
After all, I think that this function here will be called by `repair_geometry()`.

psychon commented on lib/awful/rules.lua:
```diff
@@ -363,8 +363,8 @@ function rules.extra_properties.placement(c, value)
-        honor_workarea = true,
-        honor_padding  = true
+        honor_workarea = props.honor_workarea ~= false,
+        honor_padding  = props.honor_padding ~= false
```
Undocumented?

Elv13 replied:
Nothing new in the module. I still have my `managing-clients.md.lua` doc pages, but it's incomplete. It started from the same code that ended up being `appearance.md.lua`. It parses the code (client.c and client.lua) to build a property database then add the "custom" ones.

But for now I added a new commit with a crappy header doc. It also adds all the other undocumented properties.

psychon commented on lib/awful/placement.lua:
```diff
@@ -1422,11 +1422,33 @@ function placement.restore(d, args)
+    if memento.screen and memento.screen.valid and args.context == "maximize"
+      and d.screen and get_screen(memento.screen) ~= get_screen(d.screen) then
+        -- Use the absolute geometry as the memento also does
+        local oldsgeo = memento.screen.geometry
```
Should the screen's geometry be saved inside the memento? That way this would also work if the original screen has since be removed.
Also, the screen's geometry could have changed, so the `x`,`y` saved in there need to be fixed up anyway (the `get_screen(memento.screen) ~= get_screen(d.screen) ` should be dropped), I think.

psychon commented:
Tentative :+1: besides the comments above.

Elv13 commented:
This is ready for another review and/or merge+open_issue for the remaining "problem" (that isn't really a regression)

Elv13 commented:
/me I think I will hear about these changes for some time... Hopefully they get us closer to something that works.

blueyed commented on objects/client.c:
```diff
@@ -805,9 +805,17 @@
+    CLIENT_MAXIMIZED_NONE = 0 << 0,
+    CLIENT_MAXIMIZED_V    = 1 << 0,
+    CLIENT_MAXIMIZED_H    = 1 << 1,
+    CLIENT_MAXIMIZED_BOTH = 1 << 2, /* V|H == BOTH, but ~(V|H) != ~(BOTH)... */
```
@Elv13 
The comment is not true (or I do not get it).. CLIENT_MAXIMIZED_BOTH is 4, but V|H is 3 (V being 1 and H being 0)?!
So `V|H == BOTH` is not true.

Elv13 replied:
I was referring to the concepts. When you maximize both H and V, then you are *in effect* maximized. But when you undo the operation, you are left in a potentially inconsistent state.

I agree the comment is unhelpful and misleading. I should copy/paste the [discussion comment](https://github.com/awesomeWM/awesome/pull/1453#discussion_r97938954) into the code and remove this comment.
